### ISSUE TITLE
System.IO.Packaging.PackageDigitalSignatureManager relies on obsolete Package object constructors

### ISSUE THREAD

**Issue by markholst** [dotnet/Open-XML-SDK#1487]
Hi there - we've just stumbled across this one and would appreciate your assistance.

We are using System.IO.Packaging.PackageDigitalSignatureManager, which only takes a Package object. Are you also working with the relevant team to get this upgraded?

_Originally posted by @markholst in https://github.com/dotnet/Open-XML-SDK/discussions/1414#discussioncomment-6191287_

**Comment by twsouthwick** [comment]
@markholst can you provide an example of how you use this? This package is built on top of the System.IO.Packaging space, so we can see if there's anything we need to do to support the IPackage interface

**Comment by markholst** [comment]
Hi there - relevant code snippet as below.
Very much open to a better way of doing this.

```
using System;
using System.Collections.Generic;
using System.IO.Packaging;
using System.Security.Cryptography.X509Certificates;

namespace XXX
{
    public static class DigitalSignatureHelper
    {
        private const string RtOfficeDocument = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument";

        public static void Sign(Package package, X509Certificate2 certificate)
        {
            var dsm = new PackageDigitalSignatureManager(package)
            {
                CertificateOption = CertificateEmbeddingOption.InSignaturePart
            };

            var partsToSign = new List<Uri>();
            var relationshipsToSign = new List<PackageRelationshipSelector>();
            foreach (var relationship in package.GetRelationshipsByType(RtOfficeDocument))
            {
                AddSignableItems(relationship, partsToSign, relationshipsToSign);
            }

            dsm.Sign(partsToSign, certificate, relationshipsToSign);
        }

        private static void AddSignableItems(PackageRelationship relationship, ICollection<Uri> partsToSign, ICollection<PackageRelationshipSelector> relationshipsToSign)
        {
            var selector = new PackageRelationshipSelector(relationship.SourceUri, PackageRelationshipSelectorType.Id, relationship.Id);

            relationshipsToSign.Add(selector);
            if (relationship.TargetMode != TargetMode.Internal) return;

            var part = relationship.Package.GetPart(PackUriHelper.ResolvePartUri(relationship.SourceUri, relationship.TargetUri));
            if (partsToSign.Contains(part.Uri)) return;

            partsToSign.Add(part.Uri);
            foreach (var childRelationship in part.GetRelationships())
            {
                AddSignableItems(childRelationship, partsToSign, relationshipsToSign);
            }
        }

        public static bool IsValidlySigned(Package package)
        {
            var dsm = new PackageDigitalSignatureManager(package);
            var result = dsm.VerifySignatures(true);
            return result == VerifyResult.Success;
        }
    }
}
```

**Comment by djsupermariooo** [comment]
@twsouthwick Has there been any movement on this? We're in the same situation using PackageDigitalSignatureManager in a pretty similar way as above.

**Comment by twsouthwick** [comment]
I'm confused on what you're looking from this project (or how it even fits in to your code above - there are no references to OpenXml there)

If I were in your shoes here, I would do something like the following (some pseudo code that won't compile):

```csharp
using var package = Package.Create(....);
using(var doc = WordprocessingDocument.Open(package))
{
  // work on project
}

// Using the implementation you have above
Sign(package, certificate)
```

**Comment by markholst** [comment]
Hey guys, I finally found some time and solved this. I forgot this ticket was even open.

The problem is the lack of documentation on how to achieve what is a very important and useful function! The only samples I had found used the instantiated Package class that was, until recently, exposed from the WordprocessingDocument.

I realised the Package class lives in the System.IO.Packaging namespace and can be used independently of the OpenXML library. So you instantiate this with the path of the closed office document and apply the digital signing process over this. Works a treat.
